### ISSUE TITLE
Issue with "Microsoft.Build.Packaging.Pri.Tasks.ExpandPriContent task could not be loaded"

### ISSUE THREAD

**Issue by CanePlayz** [microsoft/WindowsAppSDK#3939]
### Describe the bug

When I build my WindowsAppSDK C# project with these properties:

```XAML
    <OutputType>WinExe</OutputType>
    <TargetFramework>net8.0-windows10.0.19041.0</TargetFramework>
    <RootNamespace>RPTUHub.Desktop</RootNamespace>
    <ApplicationManifest>app.manifest</ApplicationManifest>
    <Platforms>x86;x64;arm64</Platforms>
    <RuntimeIdentifiers>win-x86;win-x64;win-arm64</RuntimeIdentifiers>
    <UseWinUI>true</UseWinUI>
    <WindowsPackaging>None</WindowsPackaging>
    <WindowsAppSDKSelfContained>true</WindowsAppSDKSelfContained>
```

I get an error. This error is also documented [here](https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/stable-channel):

> Using dotnet build with a WinAppSDK C# class library project may see a build error "Microsoft.Build.Packaging.Pri.Tasks.ExpandPriContent task could not be loaded". To resolve this issue set <EnableMsixTooling>true</EnableMsixTooling> in your project file.

It is listed under "Other limitations and known issues" under Version 1.1, so it's been around quite some time. I don't think this is a permanent limitation as the app I'm trying to build is unpackaged, so I shouldn't have to add MsixTooling anywhere.

### Steps to reproduce the bug

1. Create a WindowsAppSDK/WinUI project.
2. Try to build without `<EnableMsixTooling>true</EnableMsixTooling>` in the project properties.

### Expected behavior

Successfully build the project

### NuGet package version

Windows App SDK 1.4.2: 1.4.231008000

### Packaging type

Unpackaged

### Windows version

Insider Build (xxxxx)

### IDE

Visual Studio 2022-preview

**Comment by CanePlayz** [comment]
Sorry, seems to be working now. Apparently, there was a different issue on those runs that failed yesterday.

**Comment by CanePlayz** [comment]
Reopening this as it appeared again - this time on my Surface. I have no idea though what's the difference, using the same .NET SDK version on both systems (the .NET 8 Preview).

**Comment by DarranRowe** [comment]
It is possible that you need to reload the project if you add or remove EnableMsixTooling.
If you want to be completely unpackaged, or even have the choice to choose between packaged or unpackaged, then the other project type, Blank App, Packaged with Windows Application Packaging Project (WinUI 3 in Desktop) is a better option.

![Screenshot 2023-11-04 171020](https://github.com/microsoft/WindowsAppSDK/assets/52577874/90ca6156-5205-4d09-ba57-25188067fe21)

This is a project created with the C++ version of this template. The C# version has the same layout. If you want to be completely unpackaged then just delete the packaging project. You will still need to put `<WindowsPackageType>None</WindowsPackageType>` in the project if you want the automatic bootstrap. If you want the choice then you can do that with this layout too just by changing the startup project. The build system will build the application as fully unpackaged compatible from the main project, and then the packaging project will convert that to be packaged compatible. The only thing you would need to do is if you want to use the bootstrapper, then you must also add `<WindowsAppSDKBootstrapAutoInitializeOptions_OnPackageIdentity_NoOp>true</WindowsAppSDKBootstrapAutoInitializeOptions_OnPackageIdentity_NoOp>` in addition to WindowsPackageType.

**Comment by CanePlayz** [comment]
Thanks, I've tried out the other project template, but after removing the package project and changing nothing else, the error still occurs. That also makes sense because then it has basically become the same project as the original one. It's not a huge problem, but I think it could be fixed. But feel free to correct me on that.

Only difference between the two projects by now is that one has the (Packaged) and (Unpackaged) options in VS, while the the new one with the extra project doesn't.

![image](https://github.com/microsoft/WindowsAppSDK/assets/81419612/c7db6c5a-ae05-472b-be65-26aa75b5e344)

![image](https://github.com/microsoft/WindowsAppSDK/assets/81419612/a65aeebf-c072-4f0c-8e85-65bc23a2d5db)

The solution file is exactly the same, the project file's only difference is that one mentions assets while the other one doesn't (because the assets were in the package project).

```

Microsoft Visual Studio Solution File, Format Version 12.00
# Visual Studio Version 17
VisualStudioVersion = 17.8.34219.65
MinimumVisualStudioVersion = 10.0.40219.1
Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "RPTUHub.Desktop", "RPTUHub.Desktop\RPTUHub.Desktop.csproj", "{4965AC7F-D558-4A4A-A644-0BA701F0596F}"
EndProject
Global
	GlobalSection(SolutionConfigurationPlatforms) = preSolution
		Debug|ARM64 = Debug|ARM64
		Debug|x64 = Debug|x64
		Debug|x86 = Debug|x86
		Release|ARM64 = Release|ARM64
		Release|x64 = Release|x64
		Release|x86 = Release|x86
	EndGlobalSection
	GlobalSection(ProjectConfigurationPlatforms) = postSolution
		{4965AC7F-D558-4A4A-A644-0BA701F0596F}.Debug|ARM64.ActiveCfg = Debug|ARM64
		{4965AC7F-D558-4A4A-A644-0BA701F0596F}.Debug|ARM64.Build.0 = Debug|ARM64
		{4965AC7F-D558-4A4A-A644-0BA701F0596F}.Debug|x64.ActiveCfg = Debug|x64
		{4965AC7F-D558-4A4A-A644-0BA701F0596F}.Debug|x64.Build.0 = Debug|x64
		{4965AC7F-D558-4A4A-A644-0BA701F0596F}.Debug|x86.ActiveCfg = Debug|x86
		{4965AC7F-D558-4A4A-A644-0BA701F0596F}.Debug|x86.Build.0 = Debug|x86
		{4965AC7F-D558-4A4A-A644-0BA701F0596F}.Release|ARM64.ActiveCfg = Release|ARM64
		{4965AC7F-D558-4A4A-A644-0BA701F0596F}.Release|ARM64.Build.0 = Release|ARM64
		{4965AC7F-D558-4A4A-A644-0BA701F0596F}.Release|x64.ActiveCfg = Release|x64
		{4965AC7F-D558-4A4A-A644-0BA701F0596F}.Release|x64.Build.0 = Release|x64
		{4965AC7F-D558-4A4A-A644-0BA701F0596F}.Release|x86.ActiveCfg = Release|x86
		{4965AC7F-D558-4A4A-A644-0BA701F0596F}.Release|x86.Build.0 = Release|x86
	EndGlobalSection
	GlobalSection(SolutionProperties) = preSolution
		HideSolutionNode = FALSE
	EndGlobalSection
	GlobalSection(ExtensibilityGlobals) = postSolution
		SolutionGuid = {52F5F8D4-2701-472F-9A4B-35CAB1C4ECE3}
	EndGlobalSection
EndGlobal
```
```xaml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>WinExe</OutputType>
    <TargetFramework>net8.0-windows10.0.19041.0</TargetFramework>
    <RootNamespace>RPTUHub.Desktop</RootNamespace>
    <ApplicationManifest>app.manifest</ApplicationManifest>
    <Platforms>x86;x64;ARM64</Platforms>
    <RuntimeIdentifiers>win-x86;win-x64;win-arm64</RuntimeIdentifiers>
    <UseWinUI>true</UseWinUI>
    <WindowsPackageType>None</WindowsPackageType>
    <WindowsAppSDKSelfContained>true</WindowsAppSDKSelfContained>
    <EnableMsixTooling>true</EnableMsixTooling>
  </PropertyGroup>

  <ItemGroup>
    <Content Include="Assets\SplashScreen.scale-200.png" />
    <Content Include="Assets\LockScreenLogo.scale-200.png" />
    <Content Include="Assets\Square150x150Logo.scale-200.png" />
    <Content Include="Assets\Square44x44Logo.scale-200.png" />
    <Content Include="Assets\Square44x44Logo.targetsize-24_altform-unplated.png" />
    <Content Include="Assets\StoreLogo.png" />
    <Content Include="Assets\Wide310x150Logo.scale-200.png" />
  </ItemGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.WindowsAppSDK" Version="1.4.231008000" />
    <PackageReference Include="Microsoft.Windows.SDK.BuildTools" Version="10.0.22621.755" />
  </ItemGroup>
</Project>
```

Do you know where this difference comes from? Where is the value that makes VS behave differently?

**Comment by DarranRowe** [comment]
Well, I just don't see this problem myself. But if you want to see an example of a project which works.
For the WinUI 3 project with separate packaging project.
.csproj file
```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>WinExe</OutputType>
    <TargetFramework>net8.0-windows10.0.19041.0</TargetFramework>
    <TargetPlatformMinVersion>10.0.17763.0</TargetPlatformMinVersion>
    <RootNamespace>CsTest</RootNamespace>
    <ApplicationManifest>app.manifest</ApplicationManifest>
    <Platforms>x86;x64;ARM64</Platforms>
    <RuntimeIdentifiers>win-x86;win-x64;win-arm64</RuntimeIdentifiers>
    <UseWinUI>true</UseWinUI>
    <WindowsPackageType>None</WindowsPackageType>
    <WindowsAppSDKSelfContained>true</WindowsAppSDKSelfContained>
  </PropertyGroup>

  <ItemGroup>
    <RuntimeHostConfigurationOption Include="System.Runtime.Loader.UseRidGraph" Value="true" />
    <PackageReference Include="Microsoft.WindowsAppSDK" Version="1.4.231008000" />
    <PackageReference Include="Microsoft.Windows.SDK.BuildTools" Version="10.0.22621.2428" />
    <Manifest Include="$(ApplicationManifest)" />
  </ItemGroup>
</Project>
```
.wapproj file
```
<?xml version="1.0" encoding="utf-8"?>
<Project ToolsVersion="15.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <PropertyGroup Condition="'$(VisualStudioVersion)' == '' or '$(VisualStudioVersion)' &lt; '15.0'">
    <VisualStudioVersion>15.0</VisualStudioVersion>
  </PropertyGroup>
  <ItemGroup Label="ProjectConfigurations">
    <ProjectConfiguration Include="Debug|x86">
      <Configuration>Debug</Configuration>
      <Platform>x86</Platform>
    </ProjectConfiguration>
    <ProjectConfiguration Include="Release|x86">
      <Configuration>Release</Configuration>
      <Platform>x86</Platform>
    </ProjectConfiguration>
    <ProjectConfiguration Include="Debug|x64">
      <Configuration>Debug</Configuration>
      <Platform>x64</Platform>
    </ProjectConfiguration>
    <ProjectConfiguration Include="Release|x64">
      <Configuration>Release</Configuration>
      <Platform>x64</Platform>
    </ProjectConfiguration>
    <ProjectConfiguration Include="Debug|ARM64">
      <Configuration>Debug</Configuration>
      <Platform>ARM64</Platform>
    </ProjectConfiguration>
    <ProjectConfiguration Include="Release|ARM64">
      <Configuration>Release</Configuration>
      <Platform>ARM64</Platform>
    </ProjectConfiguration>
  </ItemGroup>
  <PropertyGroup>
    <WapProjPath Condition="'$(WapProjPath)'==''">$(MSBuildExtensionsPath)\Microsoft\DesktopBridge\</WapProjPath>
    <PathToXAMLWinRTImplementations>CsTest\</PathToXAMLWinRTImplementations>
  </PropertyGroup>
  <Import Project="$(WapProjPath)\Microsoft.DesktopBridge.props" />
  <PropertyGroup>
    <WindowsAppSDKSelfContained>true</WindowsAppSDKSelfContained>
    <ProjectGuid>c277be59-19f9-422e-94ba-9efd82954599</ProjectGuid>
    <TargetPlatformVersion>10.0.22621.0</TargetPlatformVersion>
    <TargetPlatformMinVersion>10.0.17763.0</TargetPlatformMinVersion>
    <AssetTargetFallback>net8.0-windows$(TargetPlatformVersion);$(AssetTargetFallback)</AssetTargetFallback>
    <DefaultLanguage>en-US</DefaultLanguage>
    <EntryPointProjectUniqueName>..\CsTest\CsTest.csproj</EntryPointProjectUniqueName>
  </PropertyGroup>
  <ItemGroup>
    <AppxManifest Include="Package.appxmanifest">
      <SubType>Designer</SubType>
    </AppxManifest>
  </ItemGroup>
  <ItemGroup>
    <Content Include="Images\SplashScreen.scale-200.png" />
    <Content Include="Images\LockScreenLogo.scale-200.png" />
    <Content Include="Images\Square150x150Logo.scale-200.png" />
    <Content Include="Images\Square44x44Logo.scale-200.png" />
    <Content Include="Images\Square44x44Logo.targetsize-24_altform-unplated.png" />
    <Content Include="Images\StoreLogo.png" />
    <Content Include="Images\Wide310x150Logo.scale-200.png" />
  </ItemGroup>
  <ItemGroup>
    <ProjectReference Include="..\CsTest\CsTest.csproj">
      <SkipGetTargetFrameworkProperties>True</SkipGetTargetFrameworkProperties>
      <PublishProfile>Properties\PublishProfiles\win-$(Platform).pubxml</PublishProfile>
    </ProjectReference>
  </ItemGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.WindowsAppSDK" Version="1.4.231008000">
      <IncludeAssets>build</IncludeAssets>
    </PackageReference>
    <PackageReference Include="Microsoft.Windows.SDK.BuildTools" Version="10.0.22621.2428">
      <IncludeAssets>build</IncludeAssets>
    </PackageReference>
  </ItemGroup>
  <Import Project="$(WapProjPath)\Microsoft.DesktopBridge.targets" />
</Project>
```

For the single project version.
.csproj file
```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>WinExe</OutputType>
    <TargetFramework>net8.0-windows10.0.19041.0</TargetFramework>
    <TargetPlatformMinVersion>10.0.17763.0</TargetPlatformMinVersion>
    <RootNamespace>CsTest2</RootNamespace>
    <ApplicationManifest>app.manifest</ApplicationManifest>
    <Platforms>x86;x64;ARM64</Platforms>
    <RuntimeIdentifiers>win-x86;win-x64;win-arm64</RuntimeIdentifiers>
    <PublishProfile>win-$(Platform).pubxml</PublishProfile>
    <UseWinUI>true</UseWinUI>
    <WindowsPackageType>None</WindowsPackageType>
    <WindowsAppSDKSelfContained>true</WindowsAppSDKSelfContained>
    <EnableMsixTooling>true</EnableMsixTooling>
  </PropertyGroup>

  <ItemGroup>
    <Content Include="Assets\SplashScreen.scale-200.png" />
    <Content Include="Assets\LockScreenLogo.scale-200.png" />
    <Content Include="Assets\Square150x150Logo.scale-200.png" />
    <Content Include="Assets\Square44x44Logo.scale-200.png" />
    <Content Include="Assets\Square44x44Logo.targetsize-24_altform-unplated.png" />
    <Content Include="Assets\StoreLogo.png" />
    <Content Include="Assets\Wide310x150Logo.scale-200.png" />
  </ItemGroup>

  <ItemGroup>
    <RuntimeHostConfigurationOption Include="System.Runtime.Loader.UseRidGraph" Value="true" />
    <PackageReference Include="Microsoft.WindowsAppSDK" Version="1.4.231008000" />
    <PackageReference Include="Microsoft.Windows.SDK.BuildTools" Version="10.0.22621.2428" />
    <Manifest Include="$(ApplicationManifest)" />
  </ItemGroup>

  <!-- 
    Defining the "Msix" ProjectCapability here allows the Single-project MSIX Packaging
    Tools extension to be activated for this project even if the Windows App SDK Nuget
    package has not yet been restored.
  -->
  <ItemGroup Condition="'$(DisableMsixProjectCapabilityAddedByProject)'!='true' and '$(EnableMsixTooling)'=='true'">
    <ProjectCapability Include="Msix" />
  </ItemGroup>

  <!-- 
    Defining the "HasPackageAndPublishMenuAddedByProject" property here allows the Solution 
    Explorer "Package and Publish" context menu entry to be enabled for this project even if 
    the Windows App SDK Nuget package has not yet been restored.
  -->
  <PropertyGroup Condition="'$(DisableHasPackageAndPublishMenuAddedByProject)'!='true' and '$(EnableMsixTooling)'=='true'">
    <HasPackageAndPublishMenu>true</HasPackageAndPublishMenu>
  </PropertyGroup>
</Project>
```

These project files have had the minimal amount of edits done to them to make then .NET 8 compatible. I also added the WindowsPackageType and WindowsAppSDKSelfContained since that is what you also set.
Finally, both of these had the same modifications done to the .pubxml files. They were renamed from win10-$(Platform).pubxml to win-$(Platform).pubxml. The runtime identifier was also modified. As an example, the following is the win-x86.pubxml (formally win10-x86.pubxml) file.

win-x86.pubxml
```
<?xml version="1.0" encoding="utf-8"?>
<!--
https://go.microsoft.com/fwlink/?LinkID=208121.
-->
<Project ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
	<PropertyGroup>
		<PublishProtocol>FileSystem</PublishProtocol>
		<Platform>x86</Platform>
		<RuntimeIdentifier>win-x86</RuntimeIdentifier>
		<PublishDir>bin\$(Configuration)\$(TargetFramework)\$(RuntimeIdentifier)\publish\</PublishDir>
		<SelfContained>true</SelfContained>
		<PublishSingleFile>False</PublishSingleFile>
		<PublishReadyToRun Condition="'$(Configuration)' == 'Debug'">False</PublishReadyToRun>
		<PublishReadyToRun Condition="'$(Configuration)' != 'Debug'">True</PublishReadyToRun>
		<!-- 
    See https://github.com/microsoft/CsWinRT/issues/373
    <PublishTrimmed>True</PublishTrimmed>
    -->
	</PropertyGroup>
</Project>
```

This is also using Visual Studio 2022 17.8 preview 6.

**Comment by CanePlayz** [comment]
Yeah, thanks, I will close the issue then. It was just a minor thing and since there's a known workaround and it's even mentioned in the documentation, I'm totally fine with that. And thanks for your help. Still curious why VS was showing different options, but that's not that important.

**Comment by DarranRowe** [comment]
I was unsatisfied with the answer that I previously provided, so I did a bit of prodding into the MSBuild diagnostic log files. I actually understand the differences, and have a good idea why this workaround is needed.

First, when you don't use EnableMsixTooling, the build process will use the ExpandPriContent task, and this task is backed by Microsoft.Build.Packaging.Pri.Tasks.dll. This is in the Visual Studio directory. IIRC, MSBuild still uses .NET Framework 4. There can be issues loading assemblies built for different versions of .NET into a process running a different version.
When you use EnableMsixTooling, this will redirect the makepri related tasks to the WinAppSdkGenerateProjectPriFile task. This task is backed by Microsoft.Build.Msix.dll, and can be found in the Windows App SDK NuGet package. What's more, there is a version for .NET 4.72 and a version for .NET 5.
As a complete guess, there is some functionality difference that makes the .NET 4 version incompatible with .NET 5 and newer, so when you use donet build, which I assume is using the .NET 8 runtime in your case, it is unable to use the .NET 4 assembly and requires a version explicitly for .NET 5 and newer.
I would guess that this workaround will always be needed while Visual Studio and MSBuild provides no alternative.

**Comment by CanePlayz** [comment]
Thanks a lot :)

**Comment by kythant** [comment]
Hi Folks! This issue is now fixed with our newest 1.8-preview1 release. Please check it out :)
https://www.nuget.org/packages/Microsoft.WindowsAppSDK/1.8.250814004-preview

**Comment by CanePlayz** [comment]
Thanks :)
